onnxruntime/core/common/profiler.cc

```diff
@@ -20,128 +16,8 @@ class DeviceProfiler {
   virtual ~DeviceProfiler() = default;
 };
 
-#ifdef USE_CUDA
-#define BUF_SIZE (32 * 1024)
-#define ALIGN_SIZE (8)
-#define ALIGN_BUFFER(buffer, align) \
-  (((uintptr_t)(buffer) & ((align)-1)) ? ((buffer) + (align) - ((uintptr_t)(buffer) & ((align)-1))) : (buffer))
-#define DUR(s, e) std::lround(static_cast<double>(e - s) / 1000)
-
-class CudaProfiler final: public DeviceProfiler {
- public:
-  friend class DeviceProfiler;
-  ~CudaProfiler() = default;
-  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(CudaProfiler);
-  void StartProfiling(TimePoint start_time, int pid, int tid) override;
-  std::vector<EventRecord> EndProfiling() override;
- private:
-  CudaProfiler() = default;
-  static void CUPTIAPI BufferRequested(uint8_t**, size_t*, size_t*);
-  static void CUPTIAPI BufferCompleted(CUcontext, uint32_t, uint8_t*, size_t, size_t);
-  struct KernelStat {
-    std::string name_ = {};
-    uint32_t stream_ = 0;
-    int32_t grid_x_ = 0;
-    int32_t grid_y_ = 0;
-    int32_t grid_z_ = 0;
-    int32_t block_x_ = 0;
-    int32_t block_y_ = 0;
-    int32_t block_z_ = 0;
-    int64_t start_ = 0;
-    int64_t stop_ = 0;
-  };
-  static OrtMutex mutex_;
-  static std::vector<KernelStat> stats_;
-  bool initialized_ = false;
-  TimePoint start_time_;
-  int pid_ = 0;
-  int tid_ = 0;
-  static std::atomic_flag enabled_;
-};
-
-OrtMutex CudaProfiler::mutex_;
-std::vector<CudaProfiler::KernelStat> CudaProfiler::stats_;
-std::atomic_flag CudaProfiler::enabled_;
-
-void CUPTIAPI CudaProfiler::BufferRequested(uint8_t** buffer, size_t* size, size_t* maxNumRecords) {
-  uint8_t* bfr = (uint8_t*)malloc(BUF_SIZE + ALIGN_SIZE);
-  ORT_ENFORCE(bfr, "Failed to allocate memory for cuda kernel profiling.");
-  *size = BUF_SIZE;
-  *buffer = ALIGN_BUFFER(bfr, ALIGN_SIZE);
-  *maxNumRecords = 0;
-}
-
-void CUPTIAPI CudaProfiler::BufferCompleted(CUcontext, uint32_t, uint8_t* buffer, size_t, size_t validSize) {
-  CUptiResult status;
-  CUpti_Activity* record = NULL;
-  if (validSize > 0) {
-    std::unique_lock<OrtMutex> lock(mutex_);
```

Comment:
unique_lock is for conditional variables. Here you should use lock_guard.
Besides. when possible please avoid holding a lock when you invoke external APIs.

Reply:
@snnn: thanks for the pointers. But like lock_guard, unique_lock is a general purpose mutex wrapper, it is not pertain to condition variable; And second dose this has anything to do with the deadlock? Finally, as we could see there is no "holding a lock when you invoke external APIs", what we are doing inside the block is just logging events.
Spotting a deadlock with 11.1 is good proof for us to evade the version - it might bring even more severe trouble to customers consume the release.

Reply:
https://isocpp.org/blog/2013/12/unique-lock "Use lock_guard by default especially in local scopes, and unique_lock if you need its extra features."

Reply:
We don't need to put #7632 in this release and make ORT inference packages use CUDA 11.1, but training and TRT will definitely choose CUDA 11.1(It's out of my control).

Reply:
@snnn: understand, thanks!